More reliable pixendensitydetection under Xorg/Randr

robinxan opened the pull request:
## What does the pull request do?
Physical monitor dimensions are parsed from the monitors EDID instead of requesting them from Randr.
Automatic pixeldensity detection locks to specific scalingfactors (100%, 125%,150%,175%,200%) and falls back to 100% if internal values are way off (scalingfactor higher than 300%) or no EDID is found.

## What is the current behavior?
Xrandr reports unreliable physical monitor dimensions with hacky/invalid EDIDs, e.g. no physical dimension for specific modes.
This can lead to scalingfactors of 1000% and higher.

## What is the updated/expected behavior with this PR?
- Detect the pixeldensity more reliable and don't overscale.
- Select a scalingfactor from a list of predefined scalingfactors similar to Windows (100%,125%,150%,175%.200%).

## How was the solution implemented (if it's not obvious)?
- Get the EDID via `XRRListOutputProperties` and `XRRGetOutputProperty`. [Code Reference from Chromium](https://github.com/chromium/chromium/blob/55f44515cd0b9e7739b434d1c62f4b7e321cd530/ui/base/x/x11_display_util.cc#L165-L203)
- Parse width and height (in cm) from the `Basic Display Parameters`, as these values should be more likely to be correct than the modespecific dimension.[Wikipedia Reference](https://en.wikipedia.org/wiki/Extended_Display_Identification_Data)

## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

## Breaking changes
<!--- List any breaking changes here. When the PR is merged please add an entry to https://github.com/AvaloniaUI/Avalonia/wiki/Breaking-Changes -->


## Fixed issues
Fixes #5124

robinxan committed:
More reliable pixendensitydetection under Xorg/Randr

Physical dimensions are parsed from the monitors EDID instead of Randr.
GuessPixelDensity locks to fixed values, caps at 2 and returns 1 if the supplied
dimensions are unreasonable.

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/AvaloniaUI/Avalonia?pullRequest=5178) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/AvaloniaUI/Avalonia?pullRequest=5178) before we can accept your contribution.<br/><br/>:x: robinxan [sign now](https://cla.dotnetfoundation.org/AvaloniaUI/Avalonia?pullRequest=5178)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/AvaloniaUI/Avalonia?pullRequest=5178) it.</sub>

robinxan committed:
Merge branch 'master' into master

kekekeks commented on src/Avalonia.X11/X11Screens.cs:
```diff
@@ -82,6 +82,39 @@ private void OnEvent(XEvent ev)
+            {
+                if(rrOutput == IntPtr.Zero)
+                    return null;
+                var output = Marshal.ReadInt64(rrOutput);
```
Are you sure that it's 64 bit on 32-bit ARM too? Might be better to use `ReadIntPtr`

kekekeks replied:
Actually, `_XRRMonitorInfo` is defined like this:
```cs
typedef struct _XRRMonitorInfo {
    Atom name;
    Bool primary;
    Bool automatic;
    int noutput;
    int x;
    int y;
    int width;
    int height;
    int mwidth;
    int mheight;
    RROutput *outputs;
} XRRMonitorInfo;
```
and `RROutput` is a `XID`, so we should make the `Outputs` property to be `IntPtr*`

kekekeks commented on src/Avalonia.X11/X11Screens.cs:
```diff
@@ -82,6 +82,39 @@ private void OnEvent(XEvent ev)
+                }
+                if(!hasEDID)
+                    return null;
+                XRRGetOutputProperty(_x11.Display, output, _x11.Atoms.RR_PROPERTY_RANDR_EDID, 0, 128, false, false, _x11.Atoms.AnyPropertyType, out int actualType, out int actualFormat, out int nItems, out _, out IntPtr prop);
```
Where does 128 constant comes from?

robinxan replied:
It's the length of EDID-Structure and it made sense in the code from chromium.
But after looking more into it i'm actually a bit confused.
An EDID can contain multiple extension blocks(each 128-Bytes), e.g Base + 2  Extensions => 384 Bytes.
The `length`-parameter is actually a `1/4` of the target size, so passing `128` would return the Base with 3 Extension.
We only need to pass `32` (base EDID), but i honestly don't get it.
Is the some documentation that can explain that behavior.

robinxan replied:
According to https://cgit.freedesktop.org/xorg/proto/randrproto/tree/randrproto.txt, the offset and length are multyplied by 4.
Probably because XRRGetOutputProperty can return byte-,short-, or int-Arrays.

kekekeks commented on src/Avalonia.X11/X11Screens.cs:
```diff
@@ -82,6 +82,39 @@ private void OnEvent(XEvent ev)
+                if(!hasEDID)
+                    return null;
+                XRRGetOutputProperty(_x11.Display, output, _x11.Atoms.RR_PROPERTY_RANDR_EDID, 0, 128, false, false, _x11.Atoms.AnyPropertyType, out int actualType, out int actualFormat, out int nItems, out _, out IntPtr prop);
+                if(actualType != 19) // XA_INTEGER
```
XA_INTEGER is defined in `X11Atoms` class and `Atom` enum

kekekeks commented on src/Avalonia.X11/X11Screens.cs:
```diff
@@ -82,6 +82,39 @@ private void OnEvent(XEvent ev)
+                XFree(new IntPtr(properties));
+                if(edid.Length < 22)
+                    return null;
+                var width = edid[21];
```
Since those are magic numbers, please, quote the standard, e. g.
```cs
// 15h 1 Max. Horizontal Image Size cm.
// 16h 1 Max. Vertical Image Size cm.
```

kekekeks commented on src/Avalonia.X11/X11Screens.cs:
```diff
@@ -97,24 +130,13 @@ public unsafe X11Screen[] Screens
-                                density = X11Screen.GuessPixelDensity(mon.Width, mon.MWidth);
-                        }
-
+                        var pSize = GetPhysicalMonitorSizeFromEDID(mon.Outputs);
```
I guess we should also check `NOutputs` to be `>= 1`. BTW, any idea why it can be more than 1?

robinxan replied:
This can be the case if the user creates an monitor with multiple outputs attache by running something like this:
```
$ xrandr --setmonitor DP-2-1 auto DP-2,HDMI-1
$ xrandr --listmonitors
Monitors: 1
 0: DP-2-1 5120/0x1440/1+0+0  DP-2 HDMI-1
```
So detecting the smallest scaling factor among the attached outputs should be safest way to handle this case.

kekekeks commented:
It would be nice to also get the actual monitor name from EDID too, so we could have a more fine-grained override via environment variable.

We might also define a user-wide/system-wide config and a configuration utility like Qt does.

Commit:
Changing XRRMonitorInfo.Outputs from IntPtr to IntPtr*

Change XRRMonitorInfo.Outputs from IntPtr to IntPtr* and iterate over
all Outputs to detect the smallest pixeldensity.

robinxan committed:
Merge branch 'master' into master

maxkatz6 committed:
Merge branch 'master' into master

maxkatz6 committed:
Merge branch 'master' into master
